### PR TITLE
Interleaving

### PR DESCRIPTION
**Pull request by aewallwi**
Add capability to fringe-rate filter and to average data over interleaved time sets.

### CODE REVIEW

**Comment by aewallwi** [comment]
Coverage has actually increased despite what CodeCov is claiming at the moment. Not sure why Ubuntu 3.10 is angry. Can't have everything I guess.

**Review by jsdillon — CHANGES_REQUESTED** [review]
Looking pretty good. Just a handful of requests, mostly related to documentation.

In general, I think vis_clean and its children are a total mess and this does make that a bit worse. However, I understand the desire to get this done sooner rather than later, so my concerns about vis_clean can wait for a broader overhaul.

File: hera_cal/frf.py [hunk 1]
```diff
@@ -26,10 +26,73 @@
...
 SPEED_OF_LIGHT = const.c.si.value
 SDAY_KSEC = units.sday.to("ks")
 
+def deinterleave_data_in_time(taxis, data: np.ndarray, ninterleave=1):
+    """
+    Helper function for deinterleaving *time-ordered* data along time axis.
+
+    Parameters
+    ----------
+    taxis: np.ndarray
```
[the remark this answers is not in the document]
Comment (aewallwi) [reply hera_cal/frf.py:40]: Changed.

File: hera_cal/frf.py [hunk 2]
```diff
@@ -26,10 +26,73 @@
...
+    for i in range(ninterleave):
+        tsets.append(taxis[i::ninterleave])
+        dsets.append(data[i::ninterleave])
+    
+    return tsets, dsets
+
+
+def interleave_data_in_time(dsets):
+    """
+    Helper function for interleaving sets of time-ordered data along time axis.
```
[the remark this answers is not in the document]
Comment (aewallwi) [reply hera_cal/frf.py:70]: Noted this in docstring.

File: hera_cal/frf.py [hunk 3]
```diff
@@ -26,10 +26,73 @@
...
+    dsets: list of np.ndarray
+        list of data arrays sorted into ninterleave different interleaves.
+
+    Returns
+    ----------
+    data: np.ndarray
+        ntime x nfrequency np.ndarray containing data to deinterleave. Typically complex type.
+        data is assumed to be time ordered!
+    """
+    data = []
```
[the remark this answers is not in the document]
Comment (aewallwi) [reply hera_cal/frf.py:83]: This has been done.

File: hera_cal/frf.py [hunk 4]
```diff
@@ -853,9 +916,87 @@ class FRFilter(VisClean):
...
+        ninterleave: int, optional
+            Number of interleaves to split all datacontainers into.
+            Default is 1 -> No interleaving.
+        keys: list (optional)
+            List of antpairpol keys specifying which baseline/pols to split into de-interleaved sets.
+        set_time_sets: bool, optional
+            Attach set of lits of times in each interleave set to self.
+        Returns
+        -------
+        N/A
```
[the remark this answers is not in the document]
Comment (aewallwi) [reply hera_cal/frf.py:941]: Done.

File: hera_cal/frf.py [hunk 5]
```diff
@@ -1068,6 +1230,9 @@ def tophat_frfilter(self, frate_centers, frate_half_widths, keys=None, wgts=None
             or if there is egregious cross-talk / ground pickup.
             This arg is only used if beamfitsfile is not None.
             Default is False.
+        times: np.ndarray, optional
```
[the remark this answers is not in the document]
Comment (aewallwi) [reply hera_cal/frf.py:1233]: I've changed this to seconds. In defense of KSEC, we use nanoseconds and mHz routinely for fringe-rates and delays.

File: hera_cal/frf.py [hunk 6]
```diff
@@ -1090,15 +1257,38 @@ def tophat_frfilter(self, frate_centers, frate_half_widths, keys=None, wgts=None
...
+        
         if wgts is None:
             wgts = io.DataContainer({k: (~self.flags[k]).astype(float) for k in self.flags})
         if pre_filter_modes_between_lobe_minimum_and_zero:
             self.pre_filter_resid = DataContainer({})
             self.pre_filter_resid_flags = DataContainer({})
-            filter_kwargs_no_data = copy.deepcopy(filter_kwargs)
-            if 'data' in filter_kwargs_no_data:
-                del filter_kwargs_no_data['data']
+
```
[the remark this answers is not in the document]
Comment (aewallwi) [reply hera_cal/frf.py:1266]: I definitely agree. Hopefully this is something we can look into after H4C is out the door.

File: hera_cal/frf.py [hunk 7]
```diff
@@ -1110,67 +1300,54 @@ def tophat_frfilter(self, frate_centers, frate_half_widths, keys=None, wgts=None
...
                 if not (frate_centers[k] + frate_half_widths[k] >= 0. and frate_centers[k] - frate_half_widths[k] <= 0.):
+                    opp_in_filter_kwargs = 'output_postfix' in filter_kwargs
+                    if opp_in_filter_kwargs:
+                        output_postfix = filter_kwargs.pop('output_postfix')
+                    
                     self.fourier_filter(keys=[k], filter_centers=[0.], filter_half_widths=[min_frate_abs],
-                                        mode=mode, x=self.times * SDAY_KSEC,
-                                        wgts=wgts, output_prefix='pre_filter',
+                                        mode=mode, x=times, data=input_data, flags=input_flags,
+                                        wgts=wgts, output_prefix='pre_filter', overwrite=True,
```
[the remark this answers is not in the document]
Comment (aewallwi) [reply hera_cal/frf.py:1309]: Done.

File: hera_cal/frf.py [hunk 8]
```diff
@@ -1203,31 +1380,71 @@ def time_avg_data_and_write(input_data_list, output_data, t_avg, baseline_list=N
         default is False
     flag_output: str, optional
         string to write flag output. Optional.
+    ninterleave: int, optional
+        number of subsets to break data into for interleaving.
```
[the remark this answers is not in the document]
Comment (aewallwi) [reply hera_cal/frf.py:1384]: Added documentation for this.

File: hera_cal/frf.py [hunk 9]
```diff
@@ -1284,6 +1501,10 @@ def tophat_frfilter_argparser(mode='clean'):
             "If case == 'sky': then use fringe-rates corresponding to range of ",
             "instantanous fringe-rates that include sky emission.")
     filt_options.add_argument("--case", default="sky", help=' '.join(desc), type=str)
+    desc = ("Number interleaved time subsets to split the data into ", 
+           "and apply independent fringe-rate filters. Default is 1 (no interleaved filters.")
```
[the remark this answers is not in the document]
Comment (aewallwi) [reply hera_cal/frf.py:1505]: I've added clarification of this.

File: hera_cal/frf.py [hunk 10]
```diff
@@ -1381,6 +1602,8 @@ def load_tophat_frfilter_and_write(datafile_list, case, baseline_list=None, calf
             bin fringe rates from every freq_skip channels.
             default is 1 -> takes a long time. We recommend setting this to be larger.
             only used if case == 'uvbeam'
+        ninterleave: int, optional
+            Number of interleaved sets to run time filtering on.
```
[the remark this answers is not in the document]
Comment (aewallwi) [reply hera_cal/frf.py:1606]: I've added clarification here as well.

**Comment by aewallwi** [comment]
Thanks for the review @jsdillon. I think this is ready for another look!

File: hera_cal/frf.py [hunk 11]
```diff
@@ -1382,6 +1396,13 @@ def time_avg_data_and_write(input_data_list, output_data, t_avg, baseline_list=N
         string to write flag output. Optional.
     ninterleave: int, optional
         number of subsets to break data into for interleaving.
+        this will product ninterleave different output files
```
Comment (jsdillon) [inline hera_cal/frf.py:1399]: typo: "product" vs. "produce"

**Review by jsdillon — APPROVED** [review]
Just one typo in a docstring, otherwise good to go.

File: hera_cal/frf.py (hunk 11 above)
Comment (aewallwi) [reply hera_cal/frf.py]: Fixed.

**Comment by aewallwi** [comment]
@jsdillon, sorry for the back-and-forth. Ended up adding an additional kwarg for interleaving but failed to push before you took a look. Can you give this one last pass?

File: hera_cal/frf.py [hunk 12]
```diff
@@ -1361,7 +1361,8 @@ def tophat_frfilter(self, frate_centers, frate_half_widths, keys=None, wgts=None
 def time_avg_data_and_write(input_data_list, output_data, t_avg, baseline_list=None,
                             wgt_by_nsample=True, wgt_by_favg_nsample=False, rephase=False,
                             filetype='uvh5', verbose=False, clobber=False, flag_output=None,
-                            ninterleave=1, **read_kwargs):
+                            ninterleave=1, equalize_interleave_times=True,
```
Comment (jsdillon) [inline hera_cal/frf.py:1364]: These are kinda weird, destructive choices reflecting non-idealities of hera_pspec. Seems to me like default False makes more sense...

File: hera_cal/frf.py [hunk 13]
```diff
@@ -1798,4 +1826,12 @@ def time_average_argparser():
             "will result in two output files named 'averaged_data.interleave_0.uvh5 ",
             "and 'averaged_data.interleave_1.uvh5'")
     ap.add_argument("--ninterleave", default=1, type=int, help=desc)
+    desc = ("If set to True, times of interleave files are set to actual averages of interleave sets.",
```
Comment (jsdillon) [inline hera_cal/frf.py:1829]: Same comment as above

File: scripts/time_average.py [hunk 14]
```diff
@@ -31,5 +31,7 @@
     wgt_by_nsample=not(args.dont_wgt_by_nsample),
     wgt_by_favg_nsample=args.wgt_by_favg_nsample,
     clobber=args.clobber, verbose=args.verbose
+    equalize_interleave_times=not(args.dont_equalize_interleave_times),
```
Comment (jsdillon) [inline scripts/time_average.py:34]: Same comment as above

**Review by jsdillon — CHANGES_REQUESTED** [review]
Seems basically fine, I just think making `equalize_interleave_times` and `equalize_interleave_ntimes` default True is more likely to confuse people than not.

File: scripts/time_average.py (hunk 14 above)
Comment (aewallwi) [reply scripts/time_average.py]: fixed.

File: hera_cal/frf.py (hunk 13 above)
Comment (aewallwi) [reply hera_cal/frf.py]: fixed.

File: hera_cal/frf.py (hunk 12 above)
Comment (aewallwi) [reply hera_cal/frf.py]: Done.

**Review by jsdillon — APPROVED** [review]

**Pull request merged**